<a href="https://colab.research.google.com/github/Daniel-Fernandez-951/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Daniel_Fernandez_DS20_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
!wget 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

--2020-09-08 23:53:21--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data.2’

imports-85.data.2   100%[===================>]  25.33K  --.-KB/s    in 0.03s   

2020-09-08 23:53:21 (839 KB/s) - ‘imports-85.data.2’ saved [25936/25936]



In [ ]:
import pandas as pd
import numpy as np
columns = ['symboling', 'normalized-losses', 'make', 'f-type', 'aspiration',
           'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
           'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
           'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
           'compression-ratio', 'horsepower', 'peak-rpm', 'c-mpg', 'h-mpg', 'price']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=columns)
df.head(5)

symboling normalized-losses         make  ... c-mpg h-mpg  price
0          3                 ?  alfa-romero  ...    21    27  13495
1          3                 ?  alfa-romero  ...    21    27  16500
2          1                 ?  alfa-romero  ...    19    26  16500
3          2               164         audi  ...    24    30  13950
4          2               164         audi  ...    18    22  17450

[5 rows x 26 columns]

In [ ]:
df.isnull().sum()

symboling            0
normalized-losses    0
make                 0
f-type               0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
c-mpg                0
h-mpg                0
price                0
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [ ]:
sample_30 = df.sample(n=10, random_state=30)
sample_30

symboling normalized-losses           make  ... c-mpg h-mpg  price
121          1               154       plymouth  ...    31    38   6692
176         -1                65         toyota  ...    27    32  10898
24           1               148          dodge  ...    31    38   6229
88          -1               137     mitsubishi  ...    23    30   9279
72           3               142  mercedes-benz  ...    16    18  35056
103          0               108         nissan  ...    19    25  13499
26           1               148          dodge  ...    31    38   7609
20           0                81      chevrolet  ...    38    43   6575
155          0                91         toyota  ...    27    32   8778
185          2                94     volkswagen  ...    27    34   8195

[10 rows x 26 columns]

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

# Assumptions

### H$_{\text{null}}$:
Typical fuel efficiency--of population--is not 35mpg-highway.

### H$_{\text{alt}}$:
 Highway fuel efficiency--of population--is believed to be typical at 35mpg-highway. 


In [ ]:
from scipy import stats

In [ ]:
# Apply Ttest to highway-mpg
stats.ttest_1samp(sample_30['h-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

# Answer - Practice 1-Sample T-tests (Sample Size = 10)


---
### Statistic = 0.957024
### P$_{\text{value}}$ = 0.363555
---
## Conslusion: 
Failed to reject the null hypothesis due to the P$_{\text{value}}$ > 0.05 using
the small sample size of 10.


---

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
sample_30 = df.sample(n=100, random_state=30)
stats.ttest_1samp(sample_30['h-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

# Answer - Practice 1-Sample T-tests (Sample Size = 100)


---
### Statistic = 6.5185242467
### P$_{\text{value}}$ = 0.00000000300313
---
## Conslusion: 
Suceeded in rejecting the null hypothesis due to the P$_{\text{value}}$ < 0.05 given the larger sample size being more represenative of the population as a whole. 


---

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
column = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=column)
df = df.replace({'y': 1, 'n': 0, "?": np.NaN})

In [ ]:
mask_r = df['party'] == 'republican'
mask_d = df['party'] == 'democrat'
df_r = df[mask_r]
df_d = df[mask_d]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

# Assumption (Majority Democrat Support)

### H$_{\text{null}}$:
Levels of support for `synfuels` **is not** greater among democrats than republicans.

### H$_{\text{alt}}$:
Levels of support for `synfuels` **is** greater among democrats than republicans. 

In [ ]:
# Test at 99% level
stats.ttest_ind(df_r['synfuels'], df_d['synfuels'], nan_policy='omit')

Ttest_indResult(statistic=-8.293603989407588, pvalue=1.5759322301054064e-15)

# Answer - Practice 2-Sample T-tests (Majoirty Democrat Support)


---
### Statistic = -8.293604
### P$_{\text{value}}$ = 0.0000000000000015759
---
## Conslusion: 
Suceeded in rejecting the null hypothesis due to the P$_{\text{value}}$ < 0.01 to conclude that support for `synfuels` may see higher support with democratic congress people. 
---

# Assumption (Majority Republican Support)

### H$_{\text{null}}$:
Levels of support for `water-project` **is not** greater among republicans than democrats.

### H$_{\text{alt}}$:
Levels of support for `water-project` **is** greater among republicans than democrats. 

In [ ]:
stats.ttest_ind(df_r['water-project'], df_d['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

# Answer - Practice 2-Sample T-tests (Majoirty Republican Support)


---
### Statistic = 0.08896538
### P$_{\text{value}}$ = 0.929155682
---
## Conslusion: 
Failed to reject the null hypothesis due to the P$_{\text{value}}$ > 0.01 to conclude that support for `water-project` is not over-represented by republican support. 
---

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE